packages/cli/lib/pipeline.js

```diff
@@ -2,15 +2,71 @@ import { dirname, resolve } from 'path'
 import fromStream from 'rdf-dataset-ext/fromStream.js'
 import rdf from '@zazuko/env'
 import fromFile from 'rdf-utils-fs/fromFile.js'
+import clownface from 'clownface'
 import findPipeline from '../findPipeline.js'
+import discoverManifests from './discoverManifests.js'
+import ns from './namespaces.js'
+
+const discoverOperations = async () => {
+  const ops = rdf.termMap()
+  for await (const { manifest } of discoverManifests()) {
+    manifest
+      .has(rdf.ns.rdf.type, ns.p.Operation)
+      .forEach(x => { // distinguish EcmaScript and EcmaScriptModule?
+        const link = x.out(ns.code.implementedBy).out(ns.code.link).term
+
+        ops.set(x.term, link)
+      })
+  }
+
+  return ops
+}
+
+export const desugarWith = context => dataset => {
+  const { knownOperations } = context
+  const ptr = clownface({ dataset })
+  let n = 0
+  ptr.has(ns.rdf.type, ns.p.Pipeline)
+    .out(ns.p.steps)
+    .out(ns.p.stepList)
+    .forEach(x => {
+      for (const step of x.list()) {
+        if (step.out(ns.code.implementedBy).terms.length === 0) {
+          // when there is no implementation, we expect a known operation
+          const [quad] = step.dataset.match(step.term)
+          const knownStep = knownOperations.get(quad.predicate)
+          if (knownStep) {
+            const args = step.out(quad.predicate)
+            step.deleteOut(quad.predicate)
+            // keep args only if non-empty
+            if (!ns.rdf.nil.equals(args.term)) {
+              step.addOut(ns.code.arguments, args)
+            }
+            step.addOut(ns.rdf.type, ns.p.Step)
+            const moduleNode = ptr.blankNode(`impl_${n++}`)
+            moduleNode.addOut(ns.rdf.type, ns.code.EcmaScriptModule)
+            moduleNode.addOut(ns.code.link, knownStep)
+            step.addOut(ns.code.implementedBy, moduleNode)
+          }
+        }
+      }
```

Comment:
I think I would also handle the other branch of this conditional. If you'd add the logger to the context and at least log in debug level but maybe even warn actually, as it's important that step would not be found. Also, I personally prefer less nesting

```suggestion
          if (!knownStep) {
            log.debug(`...`)
          }

          const args = step.out(quad.predicate)
          step.deleteOut(quad.predicate)
          // keep args only if non-empty
          if (!ns.rdf.nil.equals(args.term)) {
            step.addOut(ns.code.arguments, args)
          }
          step.addOut(ns.rdf.type, ns.p.Step)
          const moduleNode = ptr.blankNode(`impl_${n++}`)
          moduleNode.addOut(ns.rdf.type, ns.code.EcmaScriptModule)
          moduleNode.addOut(ns.code.link, knownStep)
          step.addOut(ns.code.implementedBy, moduleNode)  
        }
      }
```

Reply:
the step list may include a pipeline. In this case the operation is not found but we should not have a warning.
Not sure if it's worth having an exhaustive conditional logic

Reply:
I'd say it's worth it. Silently moving on will cause "strange" error. The reporting of steps which fail to load is not that great right now

With what you said in mind, you need to check that it's not in fact a pipeline and log then

Reply:
maybe we can look for operations only when the list element is neither an instance of Step nor an instance of Pipeline.
Then, we can raise a warning if the operation is not found